olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h

```diff
@@ -29,58 +30,62 @@ namespace olp {
 namespace client {
 
 /**
- * @brief Wrapper which manages cancellation state for multiple network
+ * @brief Wrapper which manages cancellation state for any asynchronous
  * operations in a thread safe way.
```

Comment:
>  thread safe way.

Because _CancellationContext_ is movable, this means that if thread#1 moves instance of _CancellationContext_ while other thread#2 executing some method, like _ExecuteOrCancelled_, this will lead to the data race (the _impl__ will be null, thus potential crash in thread#2). 

Thus, I guess, it is better instead of accessing _impl__:
```
if (!impl) {
    return;
}
```
use copy of this pointer:
```
auto impl = impl_;
if (!impl) {
    return;
}
```
Correct me If I'm wrong...

Reply:
This is a shared pimpl, using shared_ptr. Means both thread1 and thread2 should only own a shared_ptr copy of the object and not a reference or anything. If thread1 owns a copy and moves it to the next level it only moves the shared_ptr copy so the impl_ should still be valid for all. If you move to thread1 and thread2 the same instance then there is no safeguard anyway for this and you should change your code to copy capture the context.

Reply:
> Means both thread1 and thread2 should only own a shared_ptr copy of the object and not a reference or anything.

Yes, I meant the case where there is a global _CancellationContext_ instance, which can be manipulated by 2 threads simultaneously (and one of them is performing move operation). Ofc, this case is most likely will not happen in the real scenario, but, if we state that public API is thread-safe we should comply with this statement.  

And, currently, this class has thread-safe copy-ctor and copy-assignment operator. Then why not to have thread-safe move-ctor & move-assignment operator?

Reply:
But this means I need to protect the shared_ptr itself.
This is not a likely use case at all as the user only gets a copy as a return. What he does with his copy is his deal. Until now user gets this std::shared_ptr<CancellationContext> and can copy/move/crash just as easy as before.

Reply:
It doesn't matter which way user will use CancellationContext - shared_ptr, raw pointer or reference. We just need to make sure that thread-safety is guaranteed regarding its contents.